### PR TITLE
KOGITO-7133: Extend MultipleInstanceSubprocess IT test vscode-extension BPMN editor

### PR DESCRIPTION
**Pull request by zdrapela**
JIRA: https://issues.redhat.com/projects/KOGITO/issues/KOGITO-7133

* Extend MultipleInstanceSubprocess test
* Add test for UserTask On Entry/On Exit action
* Add support for offline tests

### CODE REVIEW

File: packages/vscode-extension-pack-kogito-kie-editors/package.json [hunk 1]
```diff
@@ -183,6 +183,8 @@
     "test:it": "yarn run run-script-if --bool \"$(build-env global.build.testIT)\" --then \"yarn test:it:clean\" \"cpr it-tests/resources it-tests-tmp/resources\" \"tsc --project tsconfig.it-tests.json --skipLibCheck --sourceMap true\" \"extest setup-and-run --yarn -u -e ./test-resources -o it-tests/settings.json out/*.test.js\"",
     "test:it:insider": "rimraf ./test-resources && rimraf ./out && tsc --project tsconfig.it-tests.json --skipLibCheck --sourceMap true && extest setup-and-run -t insider --yarn -u -e ./test-resources -o it-tests/settings.json out/*.test.js",
     "test:it:clean": "rimraf ./dist-it-tests && rimraf ./test-resources && rimraf ./out && rimraf ./it-tests-tmp && rimraf test-resou",
+    "test:it-offline": "yarn run run-script-if --bool \"$(build-env global.build.testIT)\" --then \"yarn test:it:clean-offline\" \"cpr it-tests/resources it-tests-tmp/resources\" \"tsc --project tsconfig.it-tests.json --skipLibCheck --sourceMap true\" \"extest setup-and-run --yarn --offline -u -e ./test-resources -o it-tests/settings.json out/*.test.js\"",
+    "test:it:clean-offline": "rimraf ./dist-it-tests && rimraf ./out && rimraf ./it-tests-tmp && rimraf test-resou",
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/package.json]: Imho we should follow the pattern with `:` so `test:it:clean:offline` and `test:it:offline` please

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts [hunk 2]
```diff
@@ -124,6 +125,50 @@ export default class PropertiesPanelHelper {
     return this;
   }
 
+  /**
+   * Select desired property value in Implementation/Execution section
+   * @param name
+   * @param value
+   */
+  public async selectImplementationExecutionValue(name: string, value: string): Promise<PropertiesPanelHelper> {
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: Not sure I understand methods name and purpose here, it is confusing. 
What if I desire to select some property value in input or checkbox?

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts [hunk 3]
```diff
@@ -124,6 +125,50 @@ export default class PropertiesPanelHelper {
...
+    return this;
+  }
+
+  /**
+   * Select desired programming language of On Entry/On Exit action.
+   * @param name
+   * @param value
+   * @returns
+   */
+  public async selectOnAction(name: string, value: string): Promise<PropertiesPanelHelper> {
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: Again, name is confusing, imho we should create a `ImplementationExecutionHelper` that will have these methods or contain helpers for widgets we test.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts [hunk 4]
```diff
@@ -340,8 +340,9 @@ describe("Editors are loading properly", () => {
     await webview.switchBack();
   });
 
-  it("Opens MultipleInstanceSubprocess.bpmn file in BPMN Editor and test value change", async function () {
+  it("Opens MultipleInstanceSubprocess.bpmn file in BPMN Editor and test Implementation/Execution value change", async function () {
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts:346]: Only thing I am missing is the Save & Reopen flow and checking that the value do not disappear (asserting them after 2nd open)

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts [hunk 5]
```diff
@@ -351,22 +352,77 @@ describe("Editors are loading properly", () => {
 
     let propertiesPanel = await bpmnEditorTester.openDiagramProperties();
 
-    const newProcessName = "Changed Multiple Instance Sub-Process";
-    await propertiesPanel.changeProperty("Name", newProcessName, "textarea");
-    await propertiesPanel.assertPropertyValue("Name", newProcessName, "textarea");
-
+    // Implementation/Execution
     await propertiesPanel.expandPropertySection(PropertiesPanelSection.IMPLEMENTATION_EXECUTION);
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts:359]: In correlation with previous comments here you would add someting like:
`const implementationExecutionHelper = await propertielPanel.expandPropertySection(...)`
`implementationExecutionHelper.changeProperty("X", value)`
`implementationExecutionHelper.assertProperty("x", X)`

For the widgets, I suggest adding them as attributes of `ImplementationExecutionHelper` and have methods like:
`getMIDataInputWidget(): MIDataInputWidget`
`getMIDataOutputWidget(): MIDataOutputWidget`
and these classes would contain:
`setMIDataInput(value)`
`setMIDataInputDataType(datatype)`
`assertMiDataInput(inputVal, dataTypeVal)`

Let me know if it does not make sense.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts [hunk 6]
```diff
@@ -143,6 +188,8 @@ export default class PropertiesPanelHelper {
    *
    * @param propertyName
    * @param propertyValue
+   * @param propertyType Type of property (select, textarea etc.). When desired property is in a widget,
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: Imho the best way to go forward is to have this method for non-widget properties and add "widgeted" properties as separate methods.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts (hunk 2 above)
Comment (domhanak) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/helpers/bpmn/PropertiesPanelHelper.ts]: Another example is what if I want to use this in testing user tasks assignments? can I use it? name suggest I can, but it does not look right.

**Review by domhanak — CHANGES_REQUESTED** [review]
Comment above, as we discussed adding object for widgets seems like the best way to go.

**Comment by tiagobento** [comment]
IT tests for the VS Code Extensions are currently disabled on our CI, as they were too unstable and were causing random build failures. See https://issues.redhat.com/browse/KOGITO-6657

This PR introduces a command that is not run as part of the CI, so no way to verify that it's working and will continue to do so.

Can you please share what's the strategy to make the new tests part of our CI checks?

**Comment by domhanak** [comment]
The tests are run nightly on our Jenkins without issues.
Command is useful for development, it will be tested.
We are focused on SWF now, there is no strategy for this.

**Comment by tiagobento** [comment]
@domhanak Thank for the info. If you're not having any issues on these tests, why not bring them back to the community CI? 
Anyway, I'm not opposing to this change, just wanted to highlight that by not running this command often, it might become outdated and loose its value.

**Comment by tiagobento** [comment]
@domhanak @zdrapela What's the status of this PR?

**Comment by zdrapela** [comment]
> @domhanak @zdrapela What's the status of this PR?

I'm working on it. Will be revised in the next few days. Sorry for the delay.

File: packages/vscode-extension-pack-kogito-kie-editors/package.json [hunk 7]
```diff
@@ -183,6 +183,8 @@
     "test:it": "pnpm run-script-if --bool \"$(build-env global.build.testIT)\" --then \"pnpm test:it:clean\" \"cpr it-tests/resources it-tests-tmp/resources\" \"tsc --project tsconfig.it-tests.json --skipLibCheck --sourceMap true\" \"extest setup-and-run --yarn -c 1.66.2 -u -e ./test-resources -o it-tests/settings.json out/*.test.js\"",
     "test:it:insider": "rimraf ./test-resources && rimraf ./out && tsc --project tsconfig.it-tests.json --skipLibCheck --sourceMap true && extest setup-and-run -t insider  --yarn -u -e ./test-resources -o it-tests/settings.json out/*.test.js",
     "test:it:clean": "rimraf ./dist-it-tests && rimraf ./test-resources && rimraf ./out && rimraf ./it-tests-tmp && rimraf test-resou && rimraf *.vsix",
+    "test:it:offline": "pnpm run-script-if --bool \"$(build-env global.build.testIT)\" --then \"pnpm test:it:clean:offline\" \"cpr it-tests/resources it-tests-tmp/resources\" \"tsc --project tsconfig.it-tests.json --skipLibCheck --sourceMap true\" \"extest setup-and-run --yarn --offline -c 1.66.2 -u -e ./test-resources -o it-tests/settings.json out/*.test.js\"",
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/package.json]: Please remove this, we will not use it much I guess.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts [hunk 8]
```diff
@@ -35,6 +35,9 @@ import { PropertiesPanelSection } from "./helpers/bpmn/PropertiesPanelHelper";
 import { TextEdit } from "vscode";
 import Correlation from "./helpers/bpmn/Correlation";
 import ProcessVariablesWidgetHelper from "./helpers/bpmn/ProcessVariablesWidgetHelper";
+import ImplementationExecutionHelper from "./helpers/bpmn/ImplementationExecutionHelper";
+import PropertiesPanelHelper from "./helpers/dmn/PropertiesPanelHelper";
+import { select } from "underscore";
```
Comment (domhanak) [inline packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts]: Are these two used somewhere?

**Review by domhanak — CHANGES_REQUESTED** [review]
One minor comment to be resolved.

File: packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts (hunk 8 above)
Comment (zdrapela) [reply packages/vscode-extension-pack-kogito-kie-editors/it-tests/extension-editors.test.ts]: They are not. Removed.

**Review by domhanak — APPROVED** [review]
Thanks, good to go once the builds are green! :green_circle:

**Review by tiagobento — APPROVED** [review]

**Pull request merged**